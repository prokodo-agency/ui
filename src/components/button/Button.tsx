"use client"
import {
  type FC,
  type MouseEvent,
  type KeyboardEvent,
  useMemo,
  useRef,
  useId,
} from "react"

import { isString } from "@/helpers/validations"

import { create } from "@/helpers/bem"

import { Icon } from "../icon"
import { Link } from "../link"
import { Loading } from "../loading"

import styles from "./Button.module.scss"

import type { ButtonDefaultProps, ButtonProps } from "./Button.model"

const bem = create(styles, "Button")

const mockEvent = {} as MouseEvent<HTMLButtonElement>
const mockKeyDownEvent = {} as KeyboardEvent<HTMLButtonElement>

export const Button: FC<ButtonProps> = ({
  id,
  fullWidth,
  color = "primary",
  iconProps = {},
  loading,
  variant = "contained",
  className,
  contentClassName,
  disabled,
  redirect,
  ...props
}) => {
  const uniqueId = useId()
  const labelRef = useRef<HTMLSpanElement | null>(null)
  const ariaLabelId = `Button-${uniqueId}`
  const isOutlined = variant === "outlined"
  const iconName = iconProps?.name
  const isIconOnly = iconName && !(props as ButtonDefaultProps)?.title
  const iconModifier = useMemo(
    () => ({
      "icon-only": Boolean(isIconOnly),
    }),
    [isIconOnly],
  )

  const renderInnerContent = () => (
    <span ref={labelRef} aria-labelledby={ariaLabelId} className={bem("title")}>
      {iconName && (
        <Icon className={bem("icon", iconModifier)} {...iconProps} />
      )}
      {(props as ButtonDefaultProps)?.title}
    </span>
  )

  const renderContent = () => {
    if (Boolean(loading)) {
      return <Loading size="xs" />
    }
    return renderInnerContent()
  }

  const renderVariant = () => {
    if (isOutlined) {
      return (
        <div className={bem("content", iconModifier, contentClassName)}>
          {renderContent()}
        </div>
      )
    }
    return renderContent()
  }

  if (redirect) {
    return (
      <Link
        disabled={disabled}
        href={redirect.href}
        className={bem(
          undefined,
          {
            "has-fullWidth": Boolean(fullWidth),
            [`has-variant-${variant}`]: true,
            [`has-bg-${color}`]: variant === "contained",
            [`has-text-${color}`]: variant === "text",
            "is-disabled": Boolean(disabled),
            ...iconModifier,
          },
          className,
        )}
        onClick={() => props?.onClick?.(mockEvent)}
        onKeyDown={() => props?.onKeyDown?.(mockKeyDownEvent)}
      >
        {renderVariant()}
      </Link>
    )
  }
  return (
    <button
      {...props}
      color={color}
      disabled={Boolean(disabled) || Boolean(loading)}
      id={`${ariaLabelId}${isString(id) ? ` ${id}` : ""}`}
      className={bem(
        undefined,
        {
          "has-fullWidth": Boolean(fullWidth),
          [`has-variant-${variant}`]: true,
          [`has-bg-${color}`]: variant === "contained",
          [`has-text-${color}`]: variant === "text",
          "is-disabled": Boolean(disabled),
          ...iconModifier,
        },
        className,
      )}
      tabIndex={
        redirect !== undefined || Boolean(disabled)
          ? -1
          : (props?.tabIndex ?? undefined)
      }
    >
      {renderVariant()}
    </button>
  )
}
