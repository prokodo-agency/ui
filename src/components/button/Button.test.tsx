import { expect } from "@jest/globals"

import { render } from "@/tests"

import { Button } from "./Button"

describe("The Button component", () => {
  it("should render Button", () => {
    const { container } = render(<Button title="Test title" />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("should render disabled Button", () => {
    const { container } = render(<Button disabled title="Test title" />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("should render loading Button", () => {
    const { container } = render(<Button loading title="Test title" />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
