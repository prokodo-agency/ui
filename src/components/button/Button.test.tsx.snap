// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Button component should render Button 1`] = `undefined`;

exports[`The Button component should render disabled Button 1`] = `undefined`;

exports[`The Button component should render loading Button 1`] = `undefined`;
